Add Kubernetes KMS API v2 support (improved)

sshcherbakov opened the pull request:
* Add KMS API v2 implementation based on previous KMS v1 implementation
* Add `--kms` command line flag to switch between KMS API versions. By default plugin starts in KMS API v2 mode. Use `--kms=v1` to revert to KMS API v1
* Upgrade Go runtime to version 1.21
* Adopt recent dependabot fixes
* Add `--key-suffix` command line flag in KMS API v2 mode to provide mechanism to safely return back to previously used key versions
* Drop vendoring
* Adopt two recent dependabot PRs by applying their dependency version upgrades:
** Bump golang.org/x/net from 0.0.0-20201021035429-f5854403a974 to 0.17.0
** Bump google.golang.org/grpc from 1.25.1 to 1.56.3

The implementation also supports automatic key rolling introduced in the Kubernetes KMS API v2.

By default plugin starts in KMS API v2 mode. Use --kms=v1 to revert to KMS API v1.

Addresses the https://github.com/GoogleCloudPlatform/k8s-cloudkms-plugin/issues/23.

Addresses the following remarks from the previous https://github.com/GoogleCloudPlatform/k8s-cloudkms-plugin/pull/27 review:
* Copyright in files left in tact where possible
* Vendoring has been dropped
* Added validation step preventing `--key-suffix` usage in `--kms=v1` mode
* Reduced code duplication in v1 and v2 implementations
* No irrelevant changes (automatically introduced by the IDE) in README file
* `errorChan` variables renamed to `errorCh` where noticed
* Adopted recommendation to avoid  forever blocking in healthz.go http.ListenAndServe() calls
* Switched both implementations to using UTC time
* Changed the `defer resp.Body.Close()` to `t.Cleanup()` in the tests
* Replaced deprectated `ioctl` with `io` package
* Adjusted plugin, plugin/v1, plugin/v2 package comments

sethvargo commented on Dockerfile:
```diff
@@ -12,7 +12,7 @@
 # limitations under the License.
 
-FROM golang:1.13 AS builder
+FROM golang:1.21 AS builder
```
Nit: Go 1.22 is out now

sethvargo commented on plugin/v1/plugin.go:
```diff
@@ -0,0 +1,109 @@
+// New constructs Plugin.
+func NewPlugin(keyService *cloudkms.ProjectsLocationsKeyRingsCryptoKeysService, keyURI, pathToUnixSocketFile string) *Plugin {
+	p := &Plugin{
+		&plugin.AbstractPlugin{
```
Don't use unnamed struct keys.

sethvargo commented on plugin/v2/healthz.go:
```diff
@@ -0,0 +1,115 @@
+	unixSocketPath string, callTimeout time.Duration, servingURL *url.URL) *HealthZ {
+
+	hz := &HealthZ{
+		&plugin.HealthZ{
```
Don't use unnamed struct keys.

sethvargo commented on plugin/v2/healthz_test.go:
```diff
@@ -0,0 +1,166 @@
+		want      int
+		keySuffix string
+	}{
+		//{
+		//	desc:     "Positive response for TestIAM, not pinging CloudKMS",
+		//	response: []json.Marshaler{positiveTestIAMResponse},
+		//	want:     http.StatusOK,
+		//},
+		{
+			desc:      "Negative response for TestIAM, not pinging CloudKMS",
+			response:  []json.Marshaler{negativeTestIAMResponse},
+			want:      http.StatusForbidden,
+			keySuffix: "test",
+		},
+		//{
+		//	desc:     "Positive response for TestIAM, Positive ping from CloudKMS",
+		//	query:    "ping-kms=true",
+		//	response: []json.Marshaler{positiveTestIAMResponse, positiveEncryptResponse, positiveDecryptResponse},
+		//	want:     http.StatusOK,
+		//},
+		//{
+		//	desc:     "Positive response for TestIAM, Negative ping from CloudKMS",
+		//	query:    "ping-kms=true",
+		//	response: []json.Marshaler{positiveTestIAMResponse, negativeEncryptResponse},
+		//	want:     http.StatusServiceUnavailable,
+		//},
+	}
```
Why are these commented out? (and above)

sethvargo commented on plugin/v2/healthz_test.go:
```diff
@@ -0,0 +1,166 @@
+			t.Parallel()
+
+			tt := setUpWithResponses(t, keyName, testCase.keySuffix, 0, testCase.response...)
+			defer tt.tearDown()
```
Prefer t.Cleanup

sethvargo commented on plugin/v2/healthz_test.go:
```diff
@@ -0,0 +1,166 @@
+	}
+
+	h := &HealthZ{
+		&plugin.HealthZ{
```
Don't use unnamed struct keys.

sethvargo commented on plugin/v2/plugin.go:
```diff
@@ -0,0 +1,158 @@
+func NewPlugin(keyService *cloudkms.ProjectsLocationsKeyRingsCryptoKeysService, keyURI, keySuffix string, pathToUnixSocketFile string) *Plugin {
+	lastKeyId = getKeyId(keyURI, keySuffix)
+	p := &Plugin{
+		&plugin.AbstractPlugin{
```
Don't use unnamed struct keys.

sethvargo commented on plugin/v2/plugin.go:
```diff
@@ -0,0 +1,158 @@
+			PathToUnixSocket: pathToUnixSocketFile,
+		},
+	}
+	p.AbstractPlugin.Plugin = p
```
Isn't this a circular reference?

sshcherbakov commented:
I have updated this PR addressing the next set of Seth's comments:

* Rebased on top of main that includes Go version upgrade to 1.22
* Replaced interface{} with any
* Only single TestIAMPermissions() method implementation left in the abstract class in plugin/healthz.go. Its signature doesn't contain context parameter, since it is not used in the pre-existing unchanged implmentation
* Adopted inlined Encrypt() call notation in PingKMS() in healthz.go files
* Removed for loop variable pinning in tests not needed in Go 1.22 any more
* Replaced defer tt.tearDown() calls with t.Cleanup(..)
* Added struct key names in the returned struct values to avoid unamed struc keys (Unnamed struc keys were used in the original implementation version)
* Several test cases were commented out in healthz_test.go for unkonwn for me reasons. I've uncommented them and fixed in the v2 version. This allowed to discover and fix a use case which I didn't anticipate: health checker doesn't pass the key resource URI to the plugin as it does Kubernetes KMS v2 API. KMS v2 plugin uses the key resource it is configured with in case if incoming key URI is empty now.

Regarding the circular references and interface embedding comments.
In order to address code duplication issue I've tried to implement an abstract class in plugin package. That is, `plugin.AbstractPlugin` contains common method implementations between v1 and v2 and not implemented methods which only have their respective implementations in the v1 and v2 packages.
For the polymorphism of an abstract class to work in Go (as I heard, this is not directly supported in the language) I used the constructions like
```
	p := &Plugin{
		AbstractPlugin: &plugin.AbstractPlugin{
                       ....
		},
	}
	p.AbstractPlugin.Plugin = p
	return p
```
which remind circular references. In fact, as understand, they are not circular references. Without `p.AbstractPlugin.Plugin = p` like the virtual table of v2.Plugin.AbstractPlugin.Plugin interface would have been filled with pointers to the base abstract class methods. `p.AbstractPlugin.Plugin = p` populates the table with pointers to the concrete method implementations and abstract class starts working. I've tried to run the code without those lines and and get SO exceptions on circular calling of empty base class interface methods.
I know abstract classes are not something Go is supposed to support. The pattern I've applied occurs in examples on the internet. Its alternatives, as I understand, include code duplication.
I am open for changing this approach. Please let me know what would be better.

sshcherbakov commented on plugin/healthz.go:
```diff
@@ -15,135 +15,110 @@
+	HandlerFunc(w http.ResponseWriter, r *http.Request)
+	NewKeyManagementServiceClient(*grpc.ClientConn) interface{}
+	PingRPC(ctx context.Context, c interface{}) error
+	TestIAMPermissions() error
```
[preexisting] TestIAMPermissions() method implementation does not use the context.

hoskeri commented on plugin/v2/plugin.go:
```diff
@@ -0,0 +1,162 @@
+	defer plugin.RecordCloudKMSOperation("encrypt", time.Now().UTC())
+
+	var response StatusResponse
+	req := &cloudkms.EncryptRequest{Plaintext: ping}
```
You probably don't want to do this on every request. 

Perhaps maintain a lastStatus + timestamp which is updated every $duration. 
You only need to trigger an explicit 'ping' if the newest response is too stale.

sshcherbakov replied:
The period for calling the Status method determines Kubernetes.
The API server polls the Status procedure call approximately every minute when everything is healthy, and every 10 seconds when the plugin is not healthy.
As I learned from Di Jin, that is what the GKE implementation doing. Using paid encrypt/decrypt call pair instead of a free get call has advantages: 1) no additional IAM permissions required 2) the status check is very accurate.
Calling Cloud KMS API twie in a minute should not be a big overhead.

hoskeri commented on plugin/v2/plugin.go:
```diff
@@ -0,0 +1,162 @@
+// to return as KeyId in case when the Cloud KMS service is not reachable
+// because KeyId in StatusResponse cannot be empty and shouldn't trigger key migration
+// in case of transient remote service unavailability
+var lastKeyId string
```
This should be be scoped to the plugin instance, and may need a lock to protect against concurrent access.

sshcherbakov replied:
There is only one process/thread updating and reading the key, as I undersand. Adding concurrency check might unnecessary complicate the implementation.

sshcherbakov replied:
Probably better to say, that concurrent access can only occur in case Kubernetes calls Status and Encrypt/Decrypt concurrently/simultaneousely. I don't think that that would work reliably on the Kubernetes side if they would be doing that. Hence I am relying here on Kubernetes tracking its KMS API calls concurency.

hoskeri commented on plugin/v2/plugin.go:
```diff
@@ -0,0 +1,162 @@
+	resp, err := g.KeyService.Encrypt(g.KeyURI, req).Context(ctx).Do()
+	if err != nil {
+		plugin.CloudKMSOperationalFailuresTotal.WithLabelValues("encrypt").Inc()
+		response = StatusResponse{Version: apiVersion, Healthz: keyNotReachable, KeyId: lastKeyId}
```
Why not return the original error here?

sshcherbakov replied:
The period for calling the Status method determines Kubernetes.
The API server polls the Status procedure call approximately every minute when everything is healthy, and every 10 seconds when the plugin is not healthy. 
As I learned from Di Jin, that is what the GKE implementation doing. Using paid encrypt/decrypt call pair instead of a free get call has advantages: 1) no additional IAM permissions required 2) the status check is very accurate.
Calling Cloud KMS API twie in a minute should not be a big overhead.
